### DISCUSSION TITLE
Enabling Support for x64dbg Debugger

### DISCUSSION THREAD

**Discussion by the-wondersmith** [Vector35/binaryninja-api#2077] [Q&A] [answered]
The more I use Binary Ninja the more enamored I become with it. To make a long post very short, I think it would be incredible if BN could use [x64dbg](https://x64dbg.com) as a back-end the way that it does gdb, lldb, and dbgeng.

Initially I looked at porting the [ret-sync](https://github.com/bootleg/ret-sync) plugin so that x64dbg and BN could be used the same way that ret-sync allows x64dbg and Ghidra to be used together. I'm still fiddling with the idea, but it's a bit above my head to figure out if going that route is the best option.

On the one hand, there appears to be a ready-made example of how it works in the form of [the existing IDA plugin](https://github.com/bootleg/ret-sync/blob/master/ext_ida/SyncPlugin.py) which is extremely convenient given that BN supports pure-python plugins. However, it's equally possible that implementing support for either [x64dbg](https://github.com/x64dbg/x64dbg) itself, [TitanEngine](https://github.com/x64dbg/TitanEngine) or [GleeBug](https://github.com/x64dbg/GleeBug) by extending BN's [DebugAdapter](https://github.com/Vector35/debugger/blob/master/DebugAdapter.py) class might be a better idea.

I have absolutely no idea, but I'm hoping someone here might. Thoughts?

**Comment by psifertex** [comment, marked as answer]
I think you're right that the best solution is port ret-sync to BN. We most certainly support pure-python plugins. In fact, all of the plugins in the current plugin manager are all python. :-) 

I can't imagine there'd be a reason you couldn't just port https://github.com/bootleg/ret-sync/blob/master/ext_ida/SyncPlugin.py and the `retsync` folder to a BN plugin that would be able to connect to a ret-sync session with another backend. 

If you do decide to take on such a task I'd recommend you join the [public slack](https://slack.binary.ninja/) and ask around in the #api-help channel, lots of folks there would be happy to help give suggestions on how to implement the different APIs.

**Reply by the-wondersmith** [reply]
@psifertex Thank you for the feedback. I've (unsurprisingly) decided to go ahead and take a crack at porting ret-sync.

My main reason for wanting a more... *native* integration with the BN Debugger is the automatic IL updating features it has. I don't suppose you have any idea if it'd be possible to make things work that way?

**Reply by psifertex** [reply]
For something like self-modifying code?

**Reply by psifertex** [reply]
Just following up again to see what you mean by automatic IL updating.

**Reply by the-wondersmith** [reply]
*SO* sorry for the delay in replying, I got a bit mired in projects.

By `automatic IL updating` I meant maintaining the same behavior being displayed in the animation under [Undiscovered Country](https://binary.ninja/2020/05/06/debugger-showcase.html#undiscovered-country) in the BN Blog post about the debugger.

The explanation text reads:

```text
If a function is purposely hidden or even just the result of a virtual function call or other indirect control flow, static analysis might miss it. But during runtime, analysis can be told the address of call destinations great improving function identification.

Prior to the call ebx, address 0x100000f25 is just data. But once this instruction is reached, Binja wisens up:
```
And the animation shows how BN intelligently re-works the IL in response to having discovered the aforementioned purposely hidden function.

**Reply by psifertex** [reply]
Ahh, yeah, gotcha. Yeah, I mean, you can totally do the same thing. The current debugger plugin is open source and it doesn't use any private APIs or anything, so you could definitely match that functionality up with an existing x64dbg session using something like retsync.
